website/blog/2022-01-31-avoid-data-silos-in-presto-in-meta.md

```diff
@@ -0,0 +1,171 @@
+
+![Benchmark](/img/blog/2022-01-31-avoid-data-silos-in-presto-in-meta/benchmark.png)
+
+As you can see from the benchmark result, P90 latency has an almost 2x improvement for RaptorX compared to Raptor. The difference between average query latency
+and P90 query latency in RaptorX is much smaller compared to Raptor. This is because in Raptor, data is physically bound to the worker node hosting it, thus a
+slow node would inevitably affect query latency. In RaptorX, instead of hard affinity between worker and data, we use soft affinity when scheduling. Soft affinity
+will select two worker nodes as candidates to process a split. If the first choice worker node is up and healthy, that node would be chosen, otherwise a secondary
+node will be chosen. Data can potentially be cached at multiple nodes, and scheduling can optimize for better CPU load balancing for the overall workloads.
+
+### Migration from Raptor to RaptorX
+All previous Raptor use cases in Meta are migrated to RaptorX, which provides better user experience and is easy to scale.
+
+#### A/B Testing Framework
+In the previous section we mentioned that the requirements for the A/B testing framework are: losslessness, flexibility, freshness, interactive latency and high availability.
+Since RaptorX is a caching layer on the original Hive data, losslessness is guaranteed by Hive. It enjoys all the query optimization from the core presto engine, as well as many
+specific optimizations in Hive connector. Benchmark shows that the query latency is comparable with Raptor, with tail latency improvements. For freshness requirements, we were
+able to benefit from Meta’s near real time warehouse data ingestion framework improvements, which improved data freshness for all Hive data. High availability was guaranteed with
+a standby cluster, same as in Raptor.
+
+During the process of migration, traffic to the framework grew to 2X due to great user experience and organic growth. RaptorX clusters were able to support the extra traffic
+with the same capacity as Raptor clusters pre-migration. The clusters’ CPU capacity were fully utilized without worrying about storage limitations.
+
+#### Dashboard
+Another typical use case of Raptor in Meta is for better dashboard experience. Presto is used to power many of the dashboarding use cases in Meta, and some data engineering
+teams choose to inject their pre-aggregated tables to dedicated Raptor clusters for better performance. By migrating to RaptorX, data engineers can remove the ingestion step
+and no longer need to worry about data consistency between base tables and the pre-aggregated tables, while also enjoying around 30% query latency reduction in most percentiles
+beyond P50.
+
+### Beyond the scope of Raptor
+Since RaptorX is very easy to use, as a booster on normal Hive connector workloads, we also enabled it for Meta's warehouse interactive workloads. These are multitenant clusters
```

Comment:
```suggestion
Since RaptorX is very easy to use as a booster on normal Hive connector workloads, we also enabled it for Meta's warehouse interactive workloads. These are multitenant clusters
```